docs/tendermint-core/p2p/peer_manager.md

```diff
@@ -0,0 +1,439 @@
+added to the peer manager.
+When the peer is reported as a `PeerStatusGood`, its score is incremented.
+When the peer is reported as a `PeerStatusBad`, its score is decremented.
+
+> The mechanisms based on the "reputation" of the peer according to reactors,
+> however, appears not to be fully implemented.
+> A peer is never `PeerStatusGood`, and is only reported as `PeerStatusBad` a
+> reactor interacting with the peer reports an error to the router, and the
+> error is not "fatal".
+> If the error is fatal, the peer is reported as [errored](#errored-peer).
+
+This score can also be _temporarily_ decremented due to connection errors.
+When the router fails to dial to a peer, it increments the peer's
+`DialFailures` counter.
+This counter is reset when the router successfully dials the peer, establishing
+a connection to it.
+During this period, between dial failures and succeeding to dial the peer, the
+peer score is decremented by the `DialFailures` counter.
+
+> `DialFailures` actually refers to a peer address. A peer may have multiple
+> addresses, and all associated counters are considered for decrementing the
+> peer's score. Also, all counters are reset when the router succeeds dialing
+> the peer.
+
+## Peer life cycle
+
+For implementing the connection policy, the peer manager keeps track of the
+state of peers and manages their life-cycle.
+The life cycle of a peer is summarized in the picture below.
+The circles represent _states_ of a peer and the rectangles represent
+_transitions_.
+All transitions are performed by the `Router`, by invoking methods of the peer
+manager with corresponding names.
+Normal transitions are represented by green arrows, while red arrows represent
+alternative transitions taken in case of errors.
+
+<img src="pics/p2p-v0.35-peermanager.png" alt="peer life cycle" title="" width="600px" name="" align="center"/>
+
+### Candidate peer
+
+The initial state of a peer in the peer manager.
+
+A `Candidate` peer may become an actual peer, to which the node is connected.
+We do not use candidate to refer to a peer to which we are connected, nor to
+a peer we are attempting to connect.
+
+Candidate peers from which the router recently disconnected or failed to dial
+are, during a certain period, not eligible for establishing connections.
+This scenario is represented by the `Frozen Candidate` state.
+
+### DialNext transition
+
+This state transition is performed when the [connection policy](#connection-policy)
+determines that the node should try to establish a connection with a peer.
+
+The peer manager selects the [best-ranked](#peer-ranking) peer which is in the
+[`Candidate`](#candidate-peer) state and provides it to the router.
+As the router is supposed to dial the peer, the peer manager sets the peer to
+the [dialing](#dialing-peer) state.
+
+Dialing a candidate peer may have become possible because the peer manager
+has found a connection slot to [upgrade](#slots-upgrades) for given room to the
```

Comment:
I will try to reword it.

We dial a peer because the policy allowed us to. This means that either: (i) we have connection slots available, or (ii) we found a peer with better score than one peer occupying a connection slot. In case (ii) we dial a peer while all connection slots are full, but we found a connection slot to upgrade. This peer is set to the upgrading sub-state, which is a mapping of two peers: the peer that we should evict and the peer that should cause the eviction. The "should" in the sentence is because we might not succeed in connecting to the higher-ranked peer; in this case, the peer to evict will not be evicted.

Reply:
Thanks for clarifying.
My initial comment was rather on the grammar of the sentence itself "[...] has found a connection slot to upgrade for given room to the [...]"

Reply:
Fixed the grammar question.